solutions/cfgym-102694C.mdx

```diff
@@ -0,0 +1,238 @@
+---
+id: cf-831C
+source: CF
+title: Sloth Naptime
+author: Sofia Yang
+---
+
+[Announcement With Official Editorial](https://codeforces.com/blog/entry/81527)
+
+**Time Complexity:** $\mathcal{O}(N \log N)$
+
+The sloth will always try to move upwards from the start node $(st)$ to the least common ancestor of the start and end nodes $(lca)$ and then move
+downwards from the least common ancestor towards the end node $(end)$.
+
+Let $dist(i, j)$ be the distance from node i to node j.
+
+So if the sloth has energy amount $e$, there are 3 separate cases:
+
+1. $e ≥ dist(st, end)$. In this case the result will be the sloth reaching the end.
+
+2. $e < dist(st, end)$ and $e < dist(st, lca)$ in which the result will be the sloth reaching the $e$th parent of $st$.
+
+3. $e < dist(st, end)$ and $e >= dist(st, lca)$ in which the result will be the sloth reaching the $(e - dist(st, lca))$th parent of $end$.
+
+So first, we can can run DFS once to find the depth of every node.
+
+Then we can create a matrix $anc$ to store the ancestors of each node,
+where $anc[i][j]$ is the $2^j$th ancestor of node $i$.
+
+After that, we can use binary lifting to answer each query by finding the least common ancestor of the start and end nodes, and then finding the right node that the sloth will reach.
+
+## Implementation
+<LanguageSection>
+
+<JavaSection>
+```java
+import java.io.*;
+import java.util.*;
+// CodeSnippet{Kattio}
```

Comment:
are you 100% sure that the codesnip goes here?

Reply:
fixed